### PR TITLE
Add HIGHLIGHT function to SQL

### LINKED ISSUE #636: Support highlighting function in SQL/PPL query engine
Related design for relevance-based search in SQL/PPL engine - https://github.com/opensearch-project/sql/issues/182

**TODO List**

- [x] Support the `highlight` function by including it in the search engine
- [x] Enable `highlight` function in SQL syntax and parser, including parameters
- [ ] Enable `highlight` function in PPL syntax and parser, including parameters
- [x] Add unit tests
- [x] Add integration tests for `highlighting`
- [x] Update the user manual - append to existing `match` search documentation

**Function Details**

(reference: https://opensearch.org/docs/latest/opensearch/ux/#highlight-query-matches)

The highlighting function adds search term results to the service response for matched terms in fields. Highlighting only works in tandem with match-search queries.  The list of highlighted fields must be present in a relevance-based search function, or else the highlighted fields will result in a syntax error. 

**Syntax**

```
highlight([field_list, …][, option=<option_value>]*) 
```

**Available Options**

- pre_tags: per-term tag to embed in the highlighted result, default is `<em>`
- post_tags: post-term tag to embed in the highlighted result, default is `</em>`

**Sample query**

```
GET shakespeare/_search
{
  "query": {
    "match": {
      "text_entry": "life"
    }
  },
  "highlight": {
    "fields": {
      "text_entry": {}
    }
  }
}
```

SQL
```
SELECT highlight("text_entry") as "highlight_text_entry" FROM shakespeare WHERE match("text_entry", "life");
```

PPL
```
source=shakespeare | where match("text_entry", "life") | highlight("text_entry")
```

**Sample response**

```
| highlight_text_entry                      |
| "my <em>life</em>, except my <em>life</em>." |
```

### PR DESCRIPTION
**Pull request by MaxKsyunz**
### Description
This PR adds a highlight function to SQL plugin. I am looking for feedback on design and implementation. 
Test coverage, code style and documentation will be addressed in the actual PR.

The design is based on how window functions are implemented.

Some questions I have,
1. At LogicalPlan level I added both `LogicalHighlight` logical plan and `HighlightExpression`. I am not sure the role `HighlightExpression` will play. Is it necessary?
2. Similarly to `WindowExpressionAnalyzer`, I created `HighlightAnanlyzer`. Its only purpose is to create  `HighlightLogical`. Is this class necessary or is it better to do this in `Analyzer.visitProject` directly?
3. Are there any scenarios to pay attention to? Any 
4. What's a good example to use to for implementing PhyiscalPlan? 

#### Sample usage
```
SELECT Tags, highlight(Tags) from Index WHERE match(Tags, 'structure')
```
would return
| Tags | highlight(Tags) |
| --- | --- |
| `data structure array` | `data <em>structure</em> array` |

### Issues Resolved
#636 

### Reviewers
@penghuo 
@joshuali925 
@dai-chen 
 
### Check List
- [ ] New functionality includes testing.
  - [ ] All tests pass, including unit test, integration test and doctest
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
  - [ ] New functionality has user manual doc added
- [ ] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: core/src/main/java/org/opensearch/sql/analysis/HighlightExpression.java [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+
+  @Override
+  public ExprType type() {
+    return ExprCoreType.STRING;
+  }
+
+  //  @Override
+  //  public <T, C> T accept(org.opensearch.sql.expression.ExpressionNodeVisitor<T, C> visitor,
+  //                         C context) {
+  //    return visitor.visitHighlight(this, context);
```
Comment (acarbonetto) [inline core/src/main/java/org/opensearch/sql/analysis/HighlightExpression.java:47]: commented code - is this a TODO?

File: opensearch/src/main/java/org/opensearch/sql/opensearch/storage/OpenSearchIndex.java [hunk 2]
```diff
@@ -120,6 +122,7 @@ public PhysicalPlan visitNode(LogicalPlan plan, OpenSearchIndexScan context) {
       }
     }
 
+
     /**
      * Implement ElasticsearchLogicalIndexScan.
```
Comment (acarbonetto) [inline opensearch/src/main/java/org/opensearch/sql/opensearch/storage/OpenSearchIndex.java:127]: Not your code, but we could update this

File: opensearch/src/main/java/org/opensearch/sql/opensearch/planner/logical/OpenSearchLogicalIndexScan.java [hunk 3]
```diff
@@ -67,14 +71,15 @@ public OpenSearchLogicalIndexScan(
       Expression filter,
       Set<ReferenceExpression> projectList,
       List<Pair<Sort.SortOption, Expression>> sortList,
-      Integer limit, Integer offset) {
+      Integer limit, Integer offset, String highlightField) {
```
Comment (penghuo) [inline opensearch/src/main/java/org/opensearch/sql/opensearch/planner/logical/OpenSearchLogicalIndexScan.java:74]: How the OpenSearchLogicalIndexScan is initialized with highlightFields?
HighlightOperator is no-op operator, which seems useless.
Have u consider just use logical expression to represent highlight fields?

File: opensearch/src/main/java/org/opensearch/sql/opensearch/executor/protector/OpenSearchExecutionProtector.java [hunk 4]
```diff
@@ -150,6 +151,14 @@ public PhysicalPlan visitAD(PhysicalPlan node, Object context) {
     );
   }
 
+  @Override
+  public PhysicalPlan visitHighlight(HighlightOperator node, Object context) {
+    return doProtect(
+        doProtect(new HighlightOperator(visitInput(node.getInput(), context),
```
Comment (Yury-Fridlyand) [inline opensearch/src/main/java/org/opensearch/sql/opensearch/executor/protector/OpenSearchExecutionProtector.java:157]: Why `doProtect` called twice?

Comment (MaxKsyunz) [reply opensearch/src/main/java/org/opensearch/sql/opensearch/executor/protector/OpenSearchExecutionProtector.java:157]: That's a typo. :/

File: opensearch/src/main/java/org/opensearch/sql/opensearch/response/OpenSearchResponse.java [hunk 5]
```diff
@@ -91,7 +92,20 @@ public Iterator<ExprValue> iterator() {
       }).iterator();
     } else {
       return Arrays.stream(hits.getHits())
-          .map(hit -> (ExprValue) exprValueFactory.construct(hit.getSourceAsString())).iterator();
+          .map(hit -> {
```
[the remark this answers is not in the document]
Comment (MaxKsyunz) [reply opensearch/src/main/java/org/opensearch/sql/opensearch/response/OpenSearchResponse.java:95]: Lucene (?) term for a match to the query.

[the remark this answers is not in the document]
Comment (acarbonetto) [reply opensearch/src/main/java/org/opensearch/sql/opensearch/response/OpenSearchResponse.java:97]: Oh: 
private final SearchHits hits;
I thought it was an abbreviation of something.  I guess that's fine, although it wouldn't hurt to call it `searchHit`.

File: opensearch/src/main/java/org/opensearch/sql/opensearch/response/OpenSearchResponse.java [hunk 6]
```diff
@@ -91,7 +94,21 @@ public Iterator<ExprValue> iterator() {
...
+            ExprValue docData = exprValueFactory.construct(hit.getSourceAsString());
+            if (hit.getHighlightFields().isEmpty()) {
+              return docData;
+            } else {
```
Comment (acarbonetto) [inline opensearch/src/main/java/org/opensearch/sql/opensearch/response/OpenSearchResponse.java:101]: else not needed. 

If we go into that branch, we return, so the else statement just adds a needless indent and statement.

File: opensearch/src/main/java/org/opensearch/sql/opensearch/response/OpenSearchResponse.java [hunk 7]
```diff
@@ -91,7 +94,21 @@ public Iterator<ExprValue> iterator() {
...
+              ImmutableMap.Builder<String, ExprValue> builder = new ImmutableMap.Builder<>();
+              builder.putAll(docData.tupleValue());
+              var hlBuilder = ImmutableMap.<String, ExprValue>builder();
```
Comment (acarbonetto) [inline opensearch/src/main/java/org/opensearch/sql/opensearch/response/OpenSearchResponse.java:104]: I personally don't like `var` (but I realize I'm in the minority on this team) but you should be consistent.  The builder above has an explicit type, and you use var here.

Comment (acarbonetto) [reply opensearch/src/main/java/org/opensearch/sql/opensearch/response/OpenSearchResponse.java:104]: Also, hlBuilder is not a great name for variable.  We don't save any money by typing fewer characters, and we shouldn't make the readers need to guess what the 'hl' stands for.  
This isn't a highlight specific class, and 'hl' isn't a known abbreviation. 
I'd recommend `highlightMapBuilder`.

File: opensearch/src/main/java/org/opensearch/sql/opensearch/response/OpenSearchResponse.java [hunk 8]
```diff
@@ -91,7 +94,21 @@ public Iterator<ExprValue> iterator() {
...
+              for (var es : hit.getHighlightFields().entrySet()) {
```
Comment (acarbonetto) [inline opensearch/src/main/java/org/opensearch/sql/opensearch/response/OpenSearchResponse.java:105]: especially for one-liners, you can consider using the lambda forEach syntax:
```
              hit.getHighlightFields().entrySet().forEach(
                      es -> hlBuilder.put(es.getKey(), ExprValueUtils.stringValue(es.getValue().toString()))
              );
```
applies elsewhere

**Review by acarbonetto — APPROVED** [review]
LG - a couple of nits

**Comment by MaxKsyunz** [comment]
Closing this draft PR since the proper one is imminent.

**Pull request closed without merging**